Parse HTTP Request message

jungcow opened the pull request:
resolves: #5, #4, #19

jungcow committed:
#5 setup: add chore test directory

jungcow committed:
#5 feat: add request parsing logic

- add startline, header, body structure
- implement startline parsing

jungcow committed:
#5 chore: add util and predefined tokenset

- add predefined method token set
- add util method for spliting string

jungcow committed:
#5 feat: add header parsing logic [prologue]

- add header buffer until header section is done
- implement logic for appending and some basic parsing header buffer

jungcow committed:
#5 feat: implement header parser logic[done]

- add rule for header field max size
- implement header parser

jungcow committed:
#5 refactor: modify body structure

- from 'std::vector<std::map<std::string, FieldValue> >'
- to 'std::vector<std::pair<Header, std::vector<char> > >'

jungcow commented:
### Testing Until Header
#### Test Request Message
```
POST / HTTP/1.1
Host: localhost:8000
User-Agent: Mozilla/5.0 (X11; Ubuntu; Linux i686; rv:29.0) Gecko/20100101 Firefox/29.0
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8
Accept-Language: en-US,en;q=0.5
Accept-Encoding: gzip, deflate
Cookie: __atuvc=34%7C7; permanent=0; _gitlab_session=226ad8a0be43681acf38c2fab9497240; __profilin=p%3Dt; request_method=GET
Connection: keep-alive
Content-Type: multipart/form-data; boundary=---------------------------9051914041544843365972754266
Content-Length: 554

-----------------------------9051914041544843365972754266
Content-Disposition: form-data; name="text"

text default
-----------------------------9051914041544843365972754266
Content-Disposition: form-data; name="file1"; filename="a.txt"
Content-Type: text/plain

Content of a.txt.

-----------------------------9051914041544843365972754266
Content-Disposition: form-data; name="file2"; filename="a.html"
Content-Type: text/html

<!DOCTYPE html><title>Content of a.html.</title>

-----------------------------9051914041544843365972754266--
```

#### Result
<img width="766" alt="Screenshot 2022-08-27 at 11 55 19 PM" src="https://user-images.githubusercontent.com/60311340/187035657-8c058b02-5e78-4594-8456-09f3312a79b8.png">
<img width="849" alt="Screenshot 2022-08-27 at 11 55 32 PM" src="https://user-images.githubusercontent.com/60311340/187035690-f389b52e-5ae5-442a-976f-bf1c008fbdb1.png">

#### 수정해야 할 것.
- [x] startline 부분의 Method가 다르다. -> 받아오는 method token의 idx가 음수가 올 수 있었으나 타입을 ssize_t가 아닌 size_t를 사용하고 있었음
- [x] Host header field에 port가 반영이 안된다. -> Header field name과 value를 `:` 문자로 split하고 split[1]을 value로 생각했었는데, value에 `:` 문자가 있는 것을 감안해야 했었다.
- [x] 희한하지만 Cookie header field에 하나가 GETGET이라는 이상한 value가 들어갔다. -> test code의 request message 중간에 `;` 문자가 빠졌더니 갯수와 실제 있는 것과의 불일치 때문에 이상한 값이 들어갔었다.
- [x] Transfer-Encoding header field에는 순서가 곧 encoding한 순서이므로 이 순서가 유지되어야 하는데, 현재는 map으로 관리하고 있기 때문에 자동으로 사전 정렬이 되어 저장된다.
- [x] fill header buffer 메소드에 들어가는 문자열은 딱 한 줄이므로 loop를 돌릴 필요가 없다.
#### 추가되어야 할 것.
- [ ] Content-Length header field의 경우 10진수만이 들어왔는지를 검사한다.

jungcow committed:
#5 feat: add contentLength body parser

- add body parser util for using content length
- add util to convert from vector to c string
- add util to convert from c string to vector

jungcow committed:
#5 feat: add chunked encoding body parser

- fix start line parsing error - not using ssize_t but using size_t
- fix header parsing error - splitting with semi colon
- fix test case error
- add chunked encoding body parser

jungcow committed:
#5 fix: fix header parser for not using loop

jungcow commented:
### Body Parsing Test
#### Test Request Message, Content-Length
```
POST / HTTP/1.1
Host: localhost:8000
User-Agent: Mozilla/5.0 (X11; Ubuntu; Linux i686; rv:29.0) Gecko/20100101 Firefox/29.0
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8
Accept-Language: en-US,en;q=0.5
Accept-Encoding: gzip, deflate
Cookie: __atuvc=34%7C7; permanent=0; _gitlab_session=226ad8a0be43681acf38c2fab9497240; __profilin=p%3Dt; request_method=GET
Connection: keep-alive
Content-Type: multipart/form-data; boundary=---------------------------9051914041544843365972754266
Content-Length: 554

-----------------------------9051914041544843365972754266
Content-Disposition: form-data; name="text"

text default
-----------------------------9051914041544843365972754266
Content-Disposition: form-data; name="file1"; filename="a.txt"
Content-Type: text/plain

Content of a.txt.

-----------------------------9051914041544843365972754266
Content-Disposition: form-data; name="file2"; filename="a.html"
Content-Type: text/html

<!DOCTYPE html><title>Content of a.html.</title>

-----------------------------9051914041544843365972754266--
```
#### Result
<img width="643" alt="Screenshot 2022-08-31 at 3 51 05 PM" src="https://user-images.githubusercontent.com/60311340/187612887-f3547e53-87ea-4308-9cf6-957f3d972228.png">

#### Test Request Message, Chunked Transfer Encoding
```
HTTP/1.1 200 OK 
Content-Type: text/plain 
Transfer-Encoding: chunked

7\r\n
Mozilla\r\n 
9\r\n
Developer\r\n
7\r\n
Network\r\n
0\r\n 
\r\n
```
#### Result
<img width="530" alt="Screenshot 2022-08-31 at 3 51 22 PM" src="https://user-images.githubusercontent.com/60311340/187612928-1cbf712c-a28d-4d61-b339-1660aa310bb2.png">

jungcow committed:
#5 chore: remove class name on util static method

jungcow commented:
### Multipart form data parsing test
<img width="483" alt="Screenshot 2022-08-31 at 7 37 50 PM" src="https://user-images.githubusercontent.com/60311340/187660143-35d8f73d-fd80-4a88-993c-6030612f8c0b.png">

jungcow committed:
#5 feat: add post body parser; multipart/form-data

seushin reviewed (approved):
구조 잡는게 쉽지 않았을 것 같은데 고생 많으셨습니다..

kimhxsong commented on webserv/src/request/request.cpp:
```diff
@@ -89,9 +89,10 @@ int Request::fillBuffer(const char *octets, size_t octetSize)
 		if (this->parseStage == Request::STAGE_BODY)
 		{
 			std::vector<char> bodyOctets(octets + parsedLength, octets + octetSize);
```
std::string 이 아닌  std::vector<char> 를 사용하는 이유가 있을까요?

jungcow replied:
null 문자를 포함한 cstr으로 string클래스를 생성할 경우 null이 잘리는 것을 발견하여 std::vector<char>로 관리를 하고자 했지만, std::string에 append메소드나 생성자에서 길이와 함께 같이 넘겨주는 방식으로 이 문제를 회피할 수 있다는 것을 늦게 깨달았습니다. 사용하는 곳이 많아 일단은 vector와 string클래스의 변환 메소드만 만들어두고 vector로 계속 관리하고자 합니다!

jungcow committed:
#5 refactor: fix cstr management to using string

jungcow committed:
Merge branch 'develop' into feature/#5

jungcow committed:
#5 refactor: add logger util, trace throw easily

jungcow committed:
#5 feat: add public method for header class

jungcow committed:
#5 feat: add static method on body class

 for converting between vec and string

jungcow commented:
logger 클래스를 이용한 throw trace와 Request message의 parsing을 통해 만들어진 데이터를 사용할 수 있도록 몇가지 interface를 만들어 두었습니다!(완료)

jungcow committed:
#5 fix: fix chunked parsing when read bytes 1 by 1

jungcow committed:
#5 refactor: fix header util and logic

- fix header methods, find method on map is not perform intentionally
- add header map type, std::less -> True, only return true
